src/transformers/models/vits/modeling_vits.py

```diff
@@ -0,0 +1,1413 @@
+            Whether or not to return a [`~utils.ModelOutput`] instead of a plain tuple.
+"""
+
+
+@add_start_docstrings(
+    "The VITS model.",
+    VITS_START_DOCSTRING,
+)
+class VitsModel(VitsPreTrainedModel):
+    def __init__(self, config: VitsConfig):
+        super().__init__(config)
+        self.config = config
+        self.text_encoder = VitsTextEncoder(config)
+        self.flow = VitsResidualCouplingBlock(config)
+        self.decoder = VitsHifiGan(config)
+
+        if config.use_stochastic_duration_prediction:
+            self.duration_predictor = VitsStochasticDurationPredictor(config)
+        else:
+            self.duration_predictor = VitsDurationPredictor(config)
+
+        if config.num_speakers > 1:
+            self.embed_speaker = nn.Embedding(config.num_speakers, config.speaker_embedding_size)
+
+        # This is used only for training.
+        self.posterior_encoder = VitsPosteriorEncoder(config)
+
+        # Initialize weights and apply final processing
+        self.post_init()
+
+    def get_encoder(self):
+        return self.text_encoder
+
+    @add_start_docstrings_to_model_forward(VITS_INPUTS_DOCSTRING)
+    @replace_return_docstrings(output_type=VitsModelOutput, config_class=_CONFIG_FOR_DOC)
+    def forward(
+        self,
+        input_ids: Optional[torch.Tensor] = None,
+        attention_mask: Optional[torch.Tensor] = None,
+        speed: float = 1.0,
+        noise_scale: float = 1.0,
+        noise_scale_duration: float = 1.0,
+        speaker_id: Optional[int] = None,
+        return_dict: Optional[bool] = None,
+        labels: Optional[torch.FloatTensor] = None,
+    ) -> Union[Tuple[torch.FloatTensor], VitsModelOutput]:
+        r"""
+        labels (`torch.FloatTensor` of shape `(batch_size, config.spectrogram_bins, sequence_length)`, *optional*):
+            Float values of target spectrogram. Timesteps set to `-100.0` are ignored (masked) for the loss
+            computation.
+
+        Returns:
+
+        Example:
+
+        ```python
+        >>> from transformers import VitsMmsTokenizer, VitsModel, set_seed
+        >>> import torch
+
+        >>> tokenizer = VitsMmsTokenizer.from_pretrained("Matthijs/mms-tts-eng")
+        >>> model = VitsModel.from_pretrained("Matthijs/mms-tts-eng")
+
+        >>> inputs = tokenizer(text="Hello, my dog is cute", return_tensors="pt")
+
+        >>> set_seed(555)  # make deterministic
+
+        >>> # generate speech
+        >>> outputs = model(inputs["input_ids"])
+        >>> outputs.audio.shape
+        torch.Size([1, 1, 33280])
+        ```
+        """
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        if attention_mask is not None:
+            input_padding_mask = attention_mask.unsqueeze(-1).float()
+        else:
+            input_padding_mask = torch.ones_like(input_ids).unsqueeze(-1).float()
+
+        # Workaround for tokenizer: filter out padding tokens.
```

Comment:
Hmm interesting - should this be reflected in the attention mask as well? Otherwise we still attend to these padding tokens no?

Reply:
They're already masked out in the attention mask. The problem with the tokenizer is that token ID 0 is used both as the padding token and as an actual character, which the HF tokenizers can't handle. So I added a real padding token to make the tokenizer happy, but the model can't handle that and so here we set those tokens to 0. But the attention mask works as usual.